File: Wire-iOS/Sources/UserInterface/Components/Views/BackgroundView/BackgroundViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "BackgroundViewController.h"
#import "UserBackgroundView.h"
#import "zmessaging+iOS.h"
@import WireExtensionComponents;

#import "UIColor+WAZExtensions.h"
#import "UIView+Borders.h"
#import "Wire-Swift.h"
#import "WAZUIMagic.h"
#import "AccentColorChangeHandler.h"
#import "Constants.h"
#import "UIView+WR_ExtendedBlockAnimations.h"
#import "ColorSchemeController.h"
#import <PureLayout/PureLayout.h>

@interface BackgroundViewController ()

@property (nonatomic) id<ZMBareUser> user;

@property (nonatomic) UIColor *filterColor;

@property (nonatomic) UserBackgroundView *backgroundView;
@property (nonatomic) NSLayoutConstraint *backgroundViewFullScreenConstraint;
@property (nonatomic) NSLayoutConstraint *backgroundViewSidebarConstraint;
@property (nonatomic) UIVisualEffectView *blurEffectView;

@property (nonatomic) AccentColorChangeHandler *accentColorHandler;

@end


@implementation BackgroundViewController

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    UIColor *accentColor = [ZMUser selfUser].accentColor;
    if (! accentColor) {
        DDLogWarn(@"User has no accent color, picking ZMAccentColorSoftPink");
        accentColor = [UIColor colorForZMAccentColor:ZMAccentColorSoftPink];
    }
    
    self.backgroundView = [[UserBackgroundView alloc] initWithFilterColor:accentColor];
    self.backgroundView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.backgroundView];

    @weakify(self);
    self.accentColorHandler = [AccentColorChangeHandler addObserver:self handlerBlock:^(UIColor *newColor, id object) {
        @strongify(self);
        self.filterColor = newColor;
    }];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(colorSchemeControllerDidApplyChanges:) name:ColorSchemeControllerDidApplyColorSchemeChangeNotification object:nil];
    
    [self.backgroundView autoPinEdgeToSuperviewEdge:ALEdgeTop];
    [self.backgroundView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
    [self.backgroundView autoPinEdgeToSuperviewEdge:ALEdgeLeft];
    
    self.backgroundViewFullScreenConstraint = [self.backgroundView autoPinEdgeToSuperviewEdge:ALEdgeRight withInset:0.0f];

    CGFloat sidebarWidth = IS_IPAD ? [WAZUIMagic cgFloatForIdentifier:@"framework.sidebar_width"] : [UIScreen mainScreen].bounds.size.width;
    self.backgroundViewSidebarConstraint = [self.backgroundView autoSetDimension:ALDimensionWidth toSize:sidebarWidth];
    [self updateBackgroundViewLayout];
    
    UIBlurEffect *blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleDark];

    self.blurEffectView = [[UIVisualEffectView alloc] initWithEffect:blurEffect];
    self.blurEffectView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.blurEffectView];
    
    [self.blurEffectView autoPinEdgesToSuperviewEdges];
}

- (void)updateBackgroundViewLayout
{
    if (self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPad) {
        self.backgroundViewFullScreenConstraint.active = NO;
        self.backgroundViewSidebarConstraint.active = YES;
    }
    else {
        self.backgroundViewSidebarConstraint.active = NO;
        self.backgroundViewFullScreenConstraint.active = YES;
    }
}

- (void)setFilterColor:(UIColor *)filterColor
{
    _filterColor = filterColor;
    self.backgroundView.filterColor = self.filterColor;
}

- (void)setUser:(id<ZMBareUser>)user animated:(BOOL)animated
{
    _user = user;
    
    [self.backgroundView setUser:user animated:animated];
}

#pragma mark - ColorSchemeControllerDidApplyChangesNotification

- (void)colorSchemeControllerDidApplyChanges:(NSNotification *)notification
{
    [self.backgroundView updateAppearanceAnimated:YES];
}

@end

File: Wire-iOS/Sources/UserInterface/ConversationList/BackgroundViewController.swift
//
// Wire
// Copyright (C) 2017 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//


import UIKit
import Cartography
import zmessaging
import WireExtensionComponents

final public class BackgroundViewController: UIViewController {
    fileprivate let imageView = UIImageView()
    private let cropView = UIView()
    private var blurView: UIVisualEffectView!
    private var userObserverToken: NSObjectProtocol! = .none
    private let user: ZMBareUser
    private let userSession: ZMUserSession?
    
    public init(user: ZMBareUser, userSession: ZMUserSession?) {
        self.user = user
        self.userSession = userSession
        super.init(nibName: .none, bundle: .none)
        
        self.userObserverToken = UserChangeInfo.add(observer: self, forBareUser: self.user)
    }
    
    public required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func viewDidLoad() {
        super.viewDidLoad()
        
        self.configureViews()
        self.createConstraints()
        
        self.updateForUser()
    }
    
    override open var prefersStatusBarHidden: Bool {
        return false
    }

    open override var preferredStatusBarStyle : UIStatusBarStyle {
        return .lightContent
    }
    
    private func configureViews() {
        self.blurView = UIVisualEffectView(effect: UIBlurEffect(style: .dark))
        
        self.cropView.clipsToBounds = true
        
        [imageView, blurView].forEach(self.cropView.addSubview)
        
        self.view.addSubview(self.cropView)
    }
    
    private func createConstraints() {
        constrain(self.view, self.imageView, self.blurView, self.cropView) { selfView, imageView, blurView, cropView in
            cropView.top == selfView.top
            cropView.bottom == selfView.bottom
            cropView.left == selfView.left - 100
            cropView.right == selfView.right + 100
            
            blurView.edges == cropView.edges
            imageView.edges == cropView.edges
        }
        
        self.blurView.transform = CGAffineTransform(scaleX: 1.4, y: 1.4)
        self.imageView.transform = CGAffineTransform(scaleX: 1.4, y: 1.4)
    }
    
    private func updateForUser() {
        guard self.isViewLoaded else {
            return
        }
        
        if user.imageMediumData == nil {
            if let searchUser = user as? ZMSearchUser, let userSession = self.userSession {
                searchUser.requestMediumProfileImage(in: userSession)
            }
            
            self.setBackground(color: user.accentColorValue.color)
        }
        else {
            self.setBackground(imageData: user.imageMediumData)
        }
    }
    
    internal func updateFor(imageMediumDataChanged: Bool, accentColorValueChanged: Bool) {
        guard imageMediumDataChanged || accentColorValueChanged else {
            return
        }
        
        if let data = user.imageMediumData, imageMediumDataChanged {
            self.setBackground(imageData: data)
        } else if accentColorValueChanged {
            self.setBackground(color: user.accentColorValue.color)
        }
    }
    
    fileprivate func setBackground(imageData: Data) {
        self.imageView.image = UIImage(data: imageData)
    }
    
    fileprivate func setBackground(color: UIColor) {
        self.imageView.image = .none
        self.imageView.backgroundColor = color
    }
}

extension BackgroundViewController: ZMUserObserver {
    public func userDidChange(_ changeInfo: UserChangeInfo) {
        self.updateFor(imageMediumDataChanged: changeInfo.imageMediumDataChanged,
                       accentColorValueChanged: changeInfo.accentColorValueChanged)
    }
}

